### ISSUE TITLE
Update optional chaining representation to match ESTree spec

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#2204]
ESTree just merged the representation for optional chaining.
https://github.com/estree/estree/pull/204

They chose a different representation to babel's representation, which we went with originally (as it was the only representation in the ecosystem).

We now need to update our parser to match this representation.

```js
interface ChainExpression <: Expression {
  type: "ChainExpression"
  expression: ChainElement
}
interface ChainElement <: Node {
  optional: boolean
}
extend interface CallExpression <: ChainElement {}
extend interface MemberExpression <: ChainElement {}
```

In terms of timelines, we need to track ESLint here, as they will be moving to implement this as well. With it, they will update their lint rules to support it.

- [x] Acorn: https://github.com/acornjs/acorn/pull/891
- [x] Espree: https://github.com/eslint/espree/pull/446
- [x] ESLint:
  - issue: https://github.com/eslint/eslint/issues/12642,
  - PR: https://github.com/eslint/eslint/pull/13416

---

This is ofc a breaking change in the AST structure, and will unfortunately be a relatively painful one for consumers of our plugins.

I don't see this as being _too_ painful, however. There will be some new false-negatives in the ecosystem due to the `ChainExpression` node, but because everything is now just a `Call/MemberExpression`, most of the plugins should "just work".

---

This is required to support ESLint 7.5.0, which added support to rules for optional chaining.
https://eslint.org/blog/2020/07/eslint-v7.5.0-released

**Comment by bradzacher** [comment]
ESLint has merged the PR adding support for the new AST.
